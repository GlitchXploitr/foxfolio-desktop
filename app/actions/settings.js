// @flow
import type { SettingsType } from '../reducers/settings';
import type { Action } from './action.d';

export function saveSettings(settings: SettingsType): Action {
  return {
    type: 'SAVE_SETTINGS',
    settings,
  };
}
