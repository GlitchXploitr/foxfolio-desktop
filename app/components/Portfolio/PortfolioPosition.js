// @flow
import type { Node } from 'react';
import React, { Component } from 'react';
import classnames from 'classnames';
import { Avatar, Card, CardContent, Grid, IconButton, Typography } from 'material-ui';
import Collapse from 'material-ui/transitions/Collapse';
import { withStyles } from 'material-ui/styles';
import { ExpandMore } from 'material-ui-icons';
import PriceChangeText from '../PriceChangeText';
import PortfolioPositionExchangeRow from './PortfolioPositionExchangeRow';
import PortfolioPositionWalletRow from './PortfolioPositionWalletRow';

export const styles = (theme: Object) => ({
  root: {
    display: 'flex',
    alignItems: 'center',
  },
  avatar: {
    flex: '0 0 auto',
    marginRight: theme.spacing.unit * 2,
  },
  content: {
    flex: '1 1 auto',
  },
  expand: {
    transform: 'rotate(0deg)',
    transition: theme.transitions.create('transform', {
      duration: theme.transitions.duration.shortest,
    }),
  },
  expandOpen: {
    transform: 'rotate(180deg)',
  },
  right: {
    textAlign: 'right',
  },
});

type Props = {
  asset: string,
  coinlist: Object,
  ticker: Object,
  portfolio: { total: number, exchanges: { [id: string]: number }, wallets: number },
  classes: any,
  fiatCurrency: string
};

type State = {
  expanded: boolean
};

class PortfolioPosition extends Component<Props, State> {
  state = {
    expanded: false,
  };

  handleExpandClick = () => {
    this.setState({ expanded: !this.state.expanded });
  };

  rowCard(avatar: Node) {
    const { asset, portfolio, classes, coinlist, ticker, fiatCurrency } = this.props;
    const quantity = portfolio.total;
    return (
      <CardContent className={classes.root} onClick={this.handleExpandClick}>
        <div className={classes.avatar}>{avatar}</div>
        <div className={classes.content}>
          <Grid container>
            <Grid item xs={3}>
              <Typography type="body2" component="span" color={quantity > 0 ? 'default' : 'secondary'}>
                {coinlist[asset].FullName}
              </Typography>
              <Typography
                type="body2"
                component="span"
                color="secondary"
              >
                {quantity.toPrecision(5)}
              </Typography>
            </Grid>
            <Grid item xs={2} className={classes.right}>
              <Typography type="body2" component="span" color={quantity > 0 ? 'default' : 'secondary'}>
                {`${(ticker[asset][fiatCurrency].PRICE * quantity).toFixed(2)}  €`}
              </Typography>
              <Typography type="body2" component="span" color={quantity > 0 ? 'default' : 'secondary'}>
                {`${(ticker[asset].BTC.PRICE * quantity).toPrecision(5)} BTC`}
              </Typography>
            </Grid>
            <Grid item xs={3} className={classes.right}>
              <Typography
                type="body2"
                component="span"
                color={quantity > 0 ? 'default' : 'secondary'}
              >
                {`${(ticker[asset][fiatCurrency].PRICE).toPrecision(5)} €`}
              </Typography>
              <Typography
                type="body2"
                component="span"
                color={quantity > 0 ? 'default' : 'secondary'}
              >
                {`${(ticker[asset].BTC.PRICE).toPrecision(5)} BTC`}
              </Typography>
            </Grid>
            <Grid item xs={2} className={classes.right}>
              <Typography
                type="body2"
                component="span"
                color={quantity > 0 ? 'default' : 'secondary'}
              >
                <PriceChangeText change={ticker[asset][fiatCurrency].CHANGEPCT24HOUR}/>
              </Typography>
              <Typography
                type="body2"
                component="span"
                color={quantity > 0 ? 'default' : 'secondary'}
              >
                <PriceChangeText change={ticker[asset].BTC.CHANGEPCT24HOUR}/>
              </Typography>
            </Grid>
          </Grid>
        </div>
        <div>
          <IconButton
            className={classnames(classes.expand, {
              [classes.expandOpen]: this.state.expanded,
            })}
          >
            <ExpandMore/>
          </IconButton>
        </div>
      </CardContent>
    );
  }

  render() {
    const { asset, coinlist, portfolio } = this.props;
    return (
      <Card>
        {this.rowCard(
          <Avatar
            src={coinlist[asset]
              ? `https://www.cryptocompare.com${coinlist[asset].ImageUrl}`
              : ''}
          />,
        )}
        <Collapse in={this.state.expanded}>
          {portfolio.wallets ? <PortfolioPositionWalletRow asset={asset} balance={portfolio.wallets}/> : ''}
          {Object.keys(portfolio.exchanges).map(key =>
            (<PortfolioPositionExchangeRow
              key={key}
              asset={asset}
              balance={portfolio.exchanges[key]}
              exchange={key.split('-')[0]}
            />))}

        </Collapse>
      </Card>
    );
  }
}

export default withStyles(styles)(PortfolioPosition);
