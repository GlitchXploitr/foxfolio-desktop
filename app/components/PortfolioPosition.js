// @flow

import type { Node } from 'react';
import React, { Component } from 'react';
import classnames from 'classnames';
import { Avatar, Card, CardContent, Grid, IconButton, Typography } from 'material-ui';
import Collapse from 'material-ui/transitions/Collapse';
import { withStyles } from 'material-ui/styles';
import { ExpandMore } from 'material-ui-icons';
import TransactionRow from './TransactionRow';

export const styles = (theme: Object) => ({
  root: {
    display: 'flex',
    alignItems: 'center',
  },
  avatar: {
    flex: '0 0 auto',
    marginRight: theme.spacing.unit * 2,
  },
  content: {
    flex: '1 1 auto',
  },
  expand: {
    transform: 'rotate(0deg)',
    transition: theme.transitions.create('transform', {
      duration: theme.transitions.duration.shortest,
    }),
  },
  expandOpen: {
    transform: 'rotate(180deg)',
  },
  right: {
    textAlign: 'right',
  },
});

type Props = {
  asset: string,
  coinlist: Object,
  ticker: Object,
  transactions: Object[],
  quantity: number,
  classes: any
};

type State = {
  expanded: boolean
};

class PortfolioPosition extends Component<Props, State> {
  state = {
    expanded: false,
  };

  handleExpandClick = () => {
    this.setState({ expanded: !this.state.expanded });
  };

  rowCard(avatar: Node) {
    const { asset, quantity, classes, coinlist, ticker } = this.props;
    return (
      <CardContent className={classes.root} onClick={this.handleExpandClick}>
        <div className={classes.avatar}>{avatar}</div>
        <div className={classes.content}>
          <Grid container>
            <Grid item xs={3}>
              <Typography type="body2" component="span">
                {coinlist[asset].FullName}
              </Typography>
              <Typography
                type="body2"
                component="span"
                color="secondary"
              >
                {quantity.toPrecision(5)}
              </Typography>
            </Grid>
            <Grid item xs={2} className={classes.right}>
              <Typography type="body2" component="span">
                {`${(ticker[asset].EUR.PRICE * quantity).toFixed(2)}  €`}
              </Typography>
              <Typography type="body2" component="span">
                {`${(ticker[asset].BTC.PRICE * quantity).toPrecision(5)} BTC`}
              </Typography>
            </Grid>
            <Grid item xs={3} className={classes.right}>
              <Typography
                type="body2"
                component="span"
                color="secondary"
              >
                {`${(ticker[asset].EUR.PRICE).toPrecision(5)} €`}
              </Typography>
              <Typography
                type="body2"
                component="span"
                color="secondary"
              >
                {`${(ticker[asset].BTC.PRICE).toPrecision(5)} BTC`}
              </Typography>
            </Grid>
            <Grid item xs={2} className={classes.right}>
              <Typography
                type="body2"
                component="span"
                color="secondary"
              >
                {`${ticker[asset].EUR.CHANGEPCT24HOUR > 0 ? '+' : ''}
                ${(ticker[asset].EUR.CHANGEPCT24HOUR).toFixed(2)}%`}
              </Typography>
              <Typography
                type="body2"
                component="span"
                color="secondary"
              >
                {`${ticker[asset].BTC.CHANGEPCT24HOUR > 0 ? '+' : ''}
                ${(ticker[asset].BTC.CHANGEPCT24HOUR).toFixed(2)}%`}
              </Typography>
            </Grid>
          </Grid>
        </div>
        <div>
          <IconButton
            className={classnames(classes.expand, {
              [classes.expandOpen]: this.state.expanded,
            })}
          >
            <ExpandMore/>
          </IconButton>
        </div>
      </CardContent>
    );
  }

  render() {
    const { asset, coinlist, ticker, transactions } = this.props;
    return (<Card>
      {this.rowCard(
        <Avatar src={coinlist[asset]
          ? `https://www.cryptocompare.com${coinlist[asset].ImageUrl}`
          : ''}
        />,
      )}
      <Collapse in={this.state.expanded}>
        {transactions.map(transaction =>
          <TransactionRow key={transaction.id} style={{ paddingLeft: 50 }} transaction={transaction} ticker={ticker}/>)}
      </Collapse>
    </Card>);
  }
}

export default withStyles(styles)(PortfolioPosition);
