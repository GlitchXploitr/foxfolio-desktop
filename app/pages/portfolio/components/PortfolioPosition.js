// @flow
import type { Node } from 'react';
import React, { Component } from 'react';
import classnames from 'classnames';
import { Card, CardContent, Grid, IconButton, Typography } from 'material-ui';
import Collapse from 'material-ui/transitions/Collapse';
import { withStyles } from 'material-ui/styles';
import { ExpandMore } from 'material-ui-icons';
import green from 'material-ui/colors/green';
import { TokenLineChart } from '../../../components/TokenLineChart';
import type { Coinlist } from '../../../reducers/coinlist/types.d';
import type { SettingsType } from '../../../reducers/settings';
import type { Ticker, TickerEntry } from '../../../reducers/ticker/types.d';

import { CurrencyAvatar } from '../../../components/CurrencyAvatar';
import PriceChangeText from './PriceChangeText';

export const styles = (theme: Object) => ({
  root: {
    display: 'flex',
    alignItems: 'center',
  },
  avatar: {
    flex: '0 0 auto',
    marginRight: theme.spacing.unit * 2,
  },
  fiatAvatar: {
    color: '#fff',
    backgroundColor: green[500],
  },
  content: {
    flex: '1 1 auto',
  },
  expand: {
    transform: 'rotate(0deg)',
    transition: theme.transitions.create('transform', {
      duration: theme.transitions.duration.shortest,
    }),
  },
  expandOpen: {
    transform: 'rotate(180deg)',
  },
  right: {
    textAlign: 'right',
  },
  collapse: {
    padding: 10,
  },
});

type Props = {
  asset: string,
  coinlist: Coinlist,
  ticker: Ticker,
  portfolio: { total: number, exchanges: { [id: string]: number }, wallets: number },
  classes: any,
  settings: SettingsType
};

type State = {
  expanded: boolean
};

type TickerForSymbol = {
  [tsym: string]: TickerEntry
};

class PortfolioPosition extends Component<Props, State> {
  state = {
    expanded: false,
  };

  handleExpandClick = () => {
    this.setState({ expanded: !this.state.expanded });
  };

  rowCard(avatar: Node) {
    const { asset, portfolio, classes, coinlist, ticker, settings } = this.props;
    const quantity = portfolio.total;

    return (
      <CardContent className={classes.root} onClick={this.handleExpandClick}>
        <div className={classes.avatar}>{avatar}</div>
        <div className={classes.content}>
          <Grid container>
            <Grid item xs={3}>
              {PositionHeader(coinlist[asset] ? coinlist[asset].FullName : asset, quantity)}
            </Grid>
            <Grid item xs={2} className={classes.right}>
              {ticker[asset] ? PositionQuantity(ticker[asset], quantity, asset, settings) : ''}
            </Grid>
            <Grid item xs={3} className={classes.right}>
              {ticker[asset] ? PositionPrice(ticker[asset], quantity, asset, settings) : ''}
            </Grid>
            <Grid item xs={2} className={classes.right}>
              {ticker[asset] ? PositionPriceChange(ticker[asset], quantity, asset, settings) : ''}
            </Grid>
          </Grid>
        </div>
        <div>
          <IconButton className={classnames(classes.expand, { [classes.expandOpen]: this.state.expanded, })}>
            <ExpandMore/>
          </IconButton>
        </div>
      </CardContent>
    );
  }

  render() {
    const { asset, classes, coinlist, settings } = this.props;
    return (
      <Card>
        {this.rowCard(<CurrencyAvatar asset={asset} coinlist={coinlist} fiatClass={classes.fiatAvatar}/>)}
        <Collapse in={this.state.expanded} timeout="auto" mountOnEnter unmountOnExit>
          <div className={classes.collapse}>
            <TokenLineChart
              fsym={asset}
              tsym={settings.currencyFocus === 'crypto' && asset !== 'BTC' ? 'BTC' : settings.fiatCurrency}
            />
          </div>
        </Collapse>
      </Card>
    );
  }
}

const PositionHeader = (name: string, quantity: number) => (
  <div>
    <Typography type="body2" component="span" color={quantity > 0 ? 'default' : 'secondary'}>
      {name}
    </Typography>
    <Typography
      type="body2"
      component="span"
      color="secondary"
    >
      {quantity.toPrecision(5)}
    </Typography>
  </div>
);

const PositionQuantity = (ticker: TickerForSymbol, quantity: number, asset: string, settings: SettingsType) => {
  const { currencyFocus, fiatCurrency } = settings;

  if (currencyFocus === 'equal') {
    return (
      <div>
        <Typography type="body2" component="span" color={quantity > 0 ? 'default' : 'secondary'}>
          {`${(parseFloat(asset !== fiatCurrency ? ticker[fiatCurrency].PRICE : 1)
            * quantity).toFixed(2)}  ${fiatCurrency}`}
        </Typography>
        <Typography type="body2" component="span" color={quantity > 0 ? 'default' : 'secondary'}>
          {`${(parseFloat(asset !== 'BTC' ? ticker.BTC.PRICE : 1) * quantity).toPrecision(5)} BTC`}
        </Typography>
      </div>
    );
  }
  const fiatEntry =
    `${(parseFloat(asset !== fiatCurrency ? ticker[fiatCurrency].PRICE : 1) * quantity).toFixed(2)}  ${fiatCurrency}`;
  const cryptoEntry = `${(parseFloat(asset !== 'BTC' ? ticker.BTC.PRICE : 1) * quantity).toPrecision(5)} BTC`;

  return (
    <div>
      <Typography type="subheading" component="span" color={quantity > 0 ? 'default' : 'secondary'}>
        {currencyFocus === 'crypto' ? cryptoEntry : fiatEntry}
      </Typography>
      <Typography type="body1" component="span" color="secondary">
        {currencyFocus === 'crypto' ? fiatEntry : cryptoEntry}
      </Typography>
    </div>
  );
};

const PositionPrice = (ticker: TickerForSymbol, quantity: number, asset: string, settings: SettingsType) => {
  const { currencyFocus, fiatCurrency } = settings;

  if (currencyFocus === 'equal') {
    return (
      <div>
        <Typography type="body2" component="span" color={quantity > 0 ? 'default' : 'secondary'}>
          {asset !== fiatCurrency ? `${parseFloat(ticker[fiatCurrency].PRICE).toPrecision(5)} ${fiatCurrency}` : '-'}
        </Typography>
        <Typography type="body2" component="span" color={quantity > 0 ? 'default' : 'secondary'}>
          {asset !== 'BTC' ? `${parseFloat(ticker.BTC.PRICE).toPrecision(5)} BTC` : '-'}
        </Typography>
      </div>
    );
  }
  const fiatEntry =
    asset !== fiatCurrency ? `${parseFloat(ticker[fiatCurrency].PRICE).toPrecision(5)} ${fiatCurrency}` : '-';
  const cryptoEntry = asset !== 'BTC' ? `${parseFloat(ticker.BTC.PRICE).toPrecision(5)} BTC` : '-';
  return (
    <div>
      <Typography type="subheading" component="span" color={quantity > 0 ? 'default' : 'secondary'}>
        {currencyFocus === 'crypto' ? cryptoEntry : fiatEntry}
      </Typography>
      <Typography type="body1" component="span" color="secondary">
        {currencyFocus === 'crypto' ? fiatEntry : cryptoEntry}
      </Typography>
    </div>
  );
};

const PositionPriceChange = (ticker: TickerForSymbol, quantity: number, asset: string, settings: SettingsType) => {
  const { currencyFocus, fiatCurrency } = settings;

  if (currencyFocus === 'equal') {
    return (
      <div>
        <Typography type="body2" component="span" color={quantity > 0 ? 'default' : 'secondary'}>
          {asset !== fiatCurrency ? <PriceChangeText change={ticker[fiatCurrency].CHANGEPCT24HOUR}/> : '-'}
        </Typography>
        <Typography type="body2" component="span" color={quantity > 0 ? 'default' : 'secondary'}>
          {asset !== 'BTC' ? <PriceChangeText change={ticker.BTC.CHANGEPCT24HOUR}/> : '-'}
        </Typography>
      </div>
    );
  }
  const fiatEntry = asset !== fiatCurrency ?
    <PriceChangeText change={ticker[fiatCurrency].CHANGEPCT24HOUR} muted={currencyFocus !== 'fiat'}/> : '-';
  const cryptoEntry = asset !== 'BTC' ?
    <PriceChangeText change={ticker.BTC.CHANGEPCT24HOUR} muted={currencyFocus !== 'crypto'}/> : '-';
  return (
    <div>
      <Typography type="subheading" component="span" color={quantity > 0 ? 'default' : 'secondary'}>
        {currencyFocus === 'crypto' ? cryptoEntry : fiatEntry}
      </Typography>
      <Typography type="body1" component="span" color="secondary">
        {currencyFocus === 'crypto' ? fiatEntry : cryptoEntry}
      </Typography>
    </div>
  );
};

export default withStyles(styles)(PortfolioPosition);
