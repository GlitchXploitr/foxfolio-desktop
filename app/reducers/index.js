// @flow
import { combineReducers } from 'redux';
import { routerReducer as router } from 'react-router-redux';
import bittrex from './bittrex';
import counter from './counter';

const rootReducer = combineReducers({
  transactions: bittrex,
  counter,
  router,
});

export default rootReducer;
