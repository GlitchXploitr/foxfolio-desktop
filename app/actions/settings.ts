import { SettingsType } from '../reducers/settings';
import { Action, Dispatch, ThunkAction } from './actions.types';
import { requestTickerUpdate } from './ticker';

function saveSettings(settings: SettingsType): Action {
  return {
    type: 'SAVE_SETTINGS',
    settings,
  };
}

export function saveSettingsAndUpdateTicker(settings: SettingsType): ThunkAction {
  return (dispatch: Dispatch) => {
    dispatch(saveSettings(settings));
    dispatch(requestTickerUpdate([], settings.fiatCurrency, settings.cryptoCurrency));
  };
}
