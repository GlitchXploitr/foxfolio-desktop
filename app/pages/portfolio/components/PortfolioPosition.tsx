import React, { Component, ComponentType } from 'react';
import classnames from 'classnames';
import { Card, CardContent, Grid, IconButton, Typography, WithStyles } from 'material-ui';
import MaterialCollapse from 'material-ui/transitions/Collapse';
import { StyleRulesCallback, Theme, withStyles } from 'material-ui/styles';
import { ExpandMore } from 'material-ui-icons';
import green from 'material-ui/colors/green';
import { CollapseProps } from 'material-ui/transitions';

import { TokenLineChart } from '../../../components/TokenLineChart';
import { Coinlist } from 'reducers/coinlist';
import { SettingsType } from '../../../reducers/settings';
import { Ticker, TickerEntry } from 'reducers/ticker';
import { CurrencyAvatar } from '../../../components/CurrencyAvatar';
import PriceChangeText from './PriceChangeText';
import { PortfolioForAsset } from './PortfolioPositions';

// TODO Remove after update to material-ui@1.0.0-beta.33
const Collapse = (MaterialCollapse as any as ComponentType<CollapseProps & { mountOnEnter: any }>);

export const styles: StyleRulesCallback = (theme: Theme) => ({
  root: {
    display: 'flex',
    alignItems: 'center',
  },
  avatar: {
    flex: '0 0 auto',
    marginRight: theme.spacing.unit * 2,
  },
  fiatAvatar: {
    color: '#fff',
    backgroundColor: green[500],
  },
  content: {
    flex: '1 1 auto',
  },
  expand: {
    transform: 'rotate(0deg)',
    transition: theme.transitions.create('transform', {
      duration: theme.transitions.duration.shortest,
    }),
  },
  expandOpen: {
    transform: 'rotate(180deg)',
  },
  right: {
    textAlign: 'right',
  },
  collapse: {
    padding: 10,
  },
});

// TODO Use connect to get coinlist, ticker and settings
type Props = {
  asset: string,
  coinlist: Coinlist,
  ticker: Ticker,
  portfolio: PortfolioForAsset,
  settings: SettingsType
};

type State = {
  expanded: boolean
};

type TickerForSymbol = {
  [tsym: string]: TickerEntry
};

export const PortfolioPosition = withStyles(styles)(
  class extends Component<Props & WithStyles, State> {
    state = {
      expanded: false,
    };

    handleExpandClick = () => {
      this.setState({ expanded: !this.state.expanded });
    };

    rowCard(avatar) {
      const { asset, portfolio, classes, coinlist, ticker, settings } = this.props;
      const quantity = portfolio.total;

      return (
        <CardContent className={classes.root} onClick={this.handleExpandClick}>
          <div className={classes.avatar}>{avatar}</div>
          <div className={classes.content}>
            <Grid container>
              <Grid item xs={3}>
                {PositionHeader(coinlist[asset] ? coinlist[asset].FullName : asset, quantity)}
              </Grid>
              <Grid item xs={2} className={classes.right}>
                {ticker[asset] ? PositionQuantity(ticker[asset], quantity, asset, settings) : ''}
              </Grid>
              <Grid item xs={3} className={classes.right}>
                {ticker[asset] ? PositionPrice(ticker[asset], quantity, asset, settings) : ''}
              </Grid>
              <Grid item xs={2} className={classes.right}>
                {ticker[asset] ? PositionPriceChange(ticker[asset], quantity, asset, settings) : ''}
              </Grid>
            </Grid>
          </div>
          <div>
            <IconButton className={classnames(classes.expand, { [classes.expandOpen]: this.state.expanded, })}>
              <ExpandMore/>
            </IconButton>
          </div>
        </CardContent>
      );
    }

    render() {
      const { asset, classes, coinlist, settings } = this.props;
      return (
        <Card>
          {this.rowCard(<CurrencyAvatar asset={asset} coinlist={coinlist} fiatClass={classes.fiatAvatar}/>)}
          <Collapse in={this.state.expanded} timeout="auto" mountOnEnter unmountOnExit>
            <div className={classes.collapse}>
              <TokenLineChart
                fsym={asset}
                tsym={settings.currencyFocus === 'crypto'
                && asset !== settings.cryptoCurrency ? settings.cryptoCurrency : settings.fiatCurrency}
              />
            </div>
          </Collapse>
        </Card>
      );
    }
  }
);

const PositionHeader = (name: string, quantity: number) => (
  <div>
    <Typography type="body2" component="span" color={quantity > 0 ? 'default' : 'secondary'}>
      {name}
    </Typography>
    <Typography
      type="body2"
      component="span"
      color="secondary"
    >
      {quantity.toPrecision(5)}
    </Typography>
  </div>
);

const PositionQuantity = (ticker: TickerForSymbol, quantity: number, asset: string, settings: SettingsType) => {
  const { currencyFocus, cryptoCurrency, fiatCurrency } = settings;

  if (currencyFocus === 'equal') {
    return (
      <div>
        <Typography type="body2" component="span" color={quantity > 0 ? 'default' : 'secondary'}>
          {`${((asset !== fiatCurrency ? parseFloat(ticker[fiatCurrency].PRICE) : 1)
            * quantity).toFixed(2)}  ${fiatCurrency}`}
        </Typography>
        <Typography type="body2" component="span" color={quantity > 0 ? 'default' : 'secondary'}>
          {`${
            ((asset !== cryptoCurrency ? parseFloat(ticker[cryptoCurrency].PRICE) : 1)
              * quantity).toPrecision(5)
            } ${cryptoCurrency}`}
        </Typography>
      </div>
    );
  }
  const fiatEntry =
    `${((asset !== fiatCurrency ? parseFloat(ticker[fiatCurrency].PRICE) : 1) * quantity).toFixed(2)}  ${fiatCurrency}`;
  const cryptoEntry =
    `${((asset !== cryptoCurrency ? parseFloat(ticker[cryptoCurrency].PRICE) : 1) * quantity).toPrecision(5)} `
    + `${cryptoCurrency}`;

  return (
    <div>
      <Typography type="subheading" component="span" color={quantity > 0 ? 'default' : 'secondary'}>
        {currencyFocus === 'crypto' ? cryptoEntry : fiatEntry}
      </Typography>
      <Typography type="body1" component="span" color="secondary">
        {currencyFocus === 'crypto' ? fiatEntry : cryptoEntry}
      </Typography>
    </div>
  );
};

const PositionPrice = (ticker: TickerForSymbol, quantity: number, asset: string, settings: SettingsType) => {
  const { currencyFocus, cryptoCurrency, fiatCurrency } = settings;

  if (currencyFocus === 'equal') {
    return (
      <div>
        <Typography type="body2" component="span" color={quantity > 0 ? 'default' : 'secondary'}>
          {asset !== fiatCurrency ? `${parseFloat(ticker[fiatCurrency].PRICE).toPrecision(5)} ${fiatCurrency}` : '-'}
        </Typography>
        <Typography type="body2" component="span" color={quantity > 0 ? 'default' : 'secondary'}>
          {asset !== cryptoCurrency
            ? `${parseFloat(ticker[cryptoCurrency].PRICE).toPrecision(5)} ${cryptoCurrency}` : '-'}
        </Typography>
      </div>
    );
  }
  const fiatEntry =
    asset !== fiatCurrency ? `${parseFloat(ticker[fiatCurrency].PRICE).toPrecision(5)} ${fiatCurrency}` : '-';
  const cryptoEntry =
    asset !== cryptoCurrency ? `${parseFloat(ticker[cryptoCurrency].PRICE).toPrecision(5)} ${cryptoCurrency}` : '-';
  return (
    <div>
      <Typography type="subheading" component="span" color={quantity > 0 ? 'default' : 'secondary'}>
        {currencyFocus === 'crypto' ? cryptoEntry : fiatEntry}
      </Typography>
      <Typography type="body1" component="span" color="secondary">
        {currencyFocus === 'crypto' ? fiatEntry : cryptoEntry}
      </Typography>
    </div>
  );
};

const PositionPriceChange = (ticker: TickerForSymbol, quantity: number, asset: string, settings: SettingsType) => {
  const { currencyFocus, cryptoCurrency, fiatCurrency } = settings;

  if (currencyFocus === 'equal') {
    return (
      <div>
        <Typography type="body2" component="span" color={quantity > 0 ? 'default' : 'secondary'}>
          {asset !== fiatCurrency ? <PriceChangeText change={ticker[fiatCurrency].CHANGEPCT24HOUR}/> : '-'}
        </Typography>
        <Typography type="body2" component="span" color={quantity > 0 ? 'default' : 'secondary'}>
          {asset !== cryptoCurrency ? <PriceChangeText change={ticker[cryptoCurrency].CHANGEPCT24HOUR}/> : '-'}
        </Typography>
      </div>
    );
  }
  const fiatEntry = asset !== fiatCurrency ?
    <PriceChangeText change={ticker[fiatCurrency].CHANGEPCT24HOUR} muted={currencyFocus !== 'fiat'}/> : '-';
  const cryptoEntry = asset !== cryptoCurrency ?
    <PriceChangeText change={ticker[cryptoCurrency].CHANGEPCT24HOUR} muted={currencyFocus !== 'crypto'}/> : '-';
  return (
    <div>
      <Typography type="subheading" component="span" color={quantity > 0 ? 'default' : 'secondary'}>
        {currencyFocus === 'crypto' ? cryptoEntry : fiatEntry}
      </Typography>
      <Typography type="body1" component="span" color="secondary">
        {currencyFocus === 'crypto' ? fiatEntry : cryptoEntry}
      </Typography>
    </div>
  );
};
