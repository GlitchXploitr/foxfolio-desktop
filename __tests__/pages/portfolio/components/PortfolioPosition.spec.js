// @flow
import { createShallow } from 'material-ui/test-utils';
import * as React from 'react';
import { initialSettings } from '../../../../app/reducers/settings';
import type { Ticker } from '../../../../app/reducers/ticker/types.d';
import PortfolioPosition from '../../../../app/pages/portfolio/components/PortfolioPosition';

let shallow;

beforeAll(() => {
  shallow = createShallow();
});

test('Render portfolio position', () => {
  const asset = 'ETH';
  const coinlist = {};
  const portfolio = {
    total: 10,
    wallets: 2,
    exchanges: { exchangeKey: 8 },
  };
  const ticker: Ticker = {
    BTC: { BTC: { PRICE: 1, CHANGEPCT24HOUR: 0 }, USD: { PRICE: 10000, CHANGEPCT24HOUR: 1.1 } },
    ETH: { BTC: { PRICE: 0.1, CHANGEPCT24HOUR: 2 }, USD: { PRICE: 1000, CHANGEPCT24HOUR: 5.2 } },
  };
  const settings = { ...initialSettings, fiatCurrency: 'USD' };

  const wrapper = shallow(
    <PortfolioPosition
      asset={asset}
      coinlist={coinlist}
      portfolio={portfolio}
      settings={settings}
      ticker={ticker}
    />).dive();
  expect(wrapper).toMatchSnapshot();
});
