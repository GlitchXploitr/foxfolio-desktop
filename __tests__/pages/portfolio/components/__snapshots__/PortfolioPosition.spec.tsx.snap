// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Portfolio position with missing ticker 1`] = `
<WithStyles(Component)
  cardContent={
    <UNDEFINED>
      <div
        className="Component-avatar-1"
      >
        <Unknown
          asset="ETH"
          coinlist={Object {}}
          fiatClass="Component-fiatAvatar-2"
        />
      </div>
      <div
        className="Component-content-3"
      >
        <WithStyles(GridWrapper)
          container={true}
        >
          <WithStyles(GridWrapper)
            item={true}
            xs={3}
          >
            <div>
              <WithStyles(Typography)
                color="default"
                component="span"
                type="body2"
              >
                ETH
              </WithStyles(Typography)>
              <WithStyles(Typography)
                color="secondary"
                component="span"
                type="body2"
              >
                10.000
              </WithStyles(Typography)>
            </div>
          </WithStyles(GridWrapper)>
          <WithStyles(GridWrapper)
            className="Component-right-4"
            item={true}
            xs={2}
          >
            
          </WithStyles(GridWrapper)>
          <WithStyles(GridWrapper)
            className="Component-right-4"
            item={true}
            xs={3}
          >
            
          </WithStyles(GridWrapper)>
          <WithStyles(GridWrapper)
            className="Component-right-4"
            item={true}
            xs={2}
          >
            
          </WithStyles(GridWrapper)>
        </WithStyles(GridWrapper)>
      </div>
    </UNDEFINED>
  }
  collapseContent={
    <Connect(WithTheme(Component))
      fsym="ETH"
      tsym="USD"
    />
  }
/>
`;

exports[`Render portfolio position 1`] = `
<WithStyles(Component)
  cardContent={
    <UNDEFINED>
      <div
        className="Component-avatar-1"
      >
        <Unknown
          asset="ETH"
          coinlist={Object {}}
          fiatClass="Component-fiatAvatar-2"
        />
      </div>
      <div
        className="Component-content-3"
      >
        <WithStyles(GridWrapper)
          container={true}
        >
          <WithStyles(GridWrapper)
            item={true}
            xs={3}
          >
            <div>
              <WithStyles(Typography)
                color="default"
                component="span"
                type="body2"
              >
                ETH
              </WithStyles(Typography)>
              <WithStyles(Typography)
                color="secondary"
                component="span"
                type="body2"
              >
                10.000
              </WithStyles(Typography)>
            </div>
          </WithStyles(GridWrapper)>
          <WithStyles(GridWrapper)
            className="Component-right-4"
            item={true}
            xs={2}
          >
            <div>
              <WithStyles(Typography)
                color="default"
                component="span"
                type="subheading"
              >
                10000.00 USD
              </WithStyles(Typography)>
              <WithStyles(Typography)
                color="secondary"
                component="span"
                type="body1"
              >
                1.0000 BTC
              </WithStyles(Typography)>
            </div>
          </WithStyles(GridWrapper)>
          <WithStyles(GridWrapper)
            className="Component-right-4"
            item={true}
            xs={3}
          >
            <div>
              <WithStyles(Typography)
                color="default"
                component="span"
                type="subheading"
              >
                1000.0 USD
              </WithStyles(Typography)>
              <WithStyles(Typography)
                color="secondary"
                component="span"
                type="body1"
              >
                0.10000 BTC
              </WithStyles(Typography)>
            </div>
          </WithStyles(GridWrapper)>
          <WithStyles(GridWrapper)
            className="Component-right-4"
            item={true}
            xs={2}
          >
            <div>
              <WithStyles(Typography)
                color="default"
                component="span"
                type="subheading"
              >
                <WithStyles(Component)
                  change={5.2}
                  muted={false}
                />
              </WithStyles(Typography)>
              <WithStyles(Typography)
                color="secondary"
                component="span"
                type="body1"
              >
                <WithStyles(Component)
                  change={2}
                  muted={true}
                />
              </WithStyles(Typography)>
            </div>
          </WithStyles(GridWrapper)>
        </WithStyles(GridWrapper)>
      </div>
    </UNDEFINED>
  }
  collapseContent={
    <Connect(WithTheme(Component))
      fsym="ETH"
      tsym="USD"
    />
  }
/>
`;
